Extensions/3D/JsExtension.js

```diff
@@ -2316,36 +2322,192 @@ module.exports = {
+        );
+
+        // Stretch the model in a 1x1x1 cube.
+        const scaleX = 1 / modelWidth;
+        const scaleY = 1 / modelHeight;
+        const scaleZ = 1 / modelDepth;
+
+        const scaleMatrix = new THREE.Matrix4();
+        // Flip on Y because the Y axis is on the opposite side of direct basis.
+        // It avoids models to be like a mirror refection.
+        scaleMatrix.makeScale(scaleX, -scaleY, scaleZ);
+        threeObject.updateMatrix();
+        threeObject.applyMatrix4(scaleMatrix);
+
+        if (keepAspectRatio) {
+          // Reduce the object dimensions to keep aspect ratio.
+          const widthRatio = originalWidth / modelWidth;
+          const heightRatio = originalHeight / modelHeight;
+          const depthRatio = originalDepth / modelDepth;
+          const scaleRatio = Math.min(widthRatio, heightRatio, depthRatio);
+
+          this._defaultWidth = scaleRatio * modelWidth;
+          this._defaultHeight = scaleRatio * modelHeight;
+          this._defaultDepth = scaleRatio * modelDepth;
+        }
+      }
+
       update() {
         const width = this.getWidth();
         const height = this.getHeight();
+        const centerPoint = this.getCenterPoint();
+        const centerX = width * centerPoint[0];
+        const centerY = height * centerPoint[1];
 
+        const minX = 0 - centerX;
+        const minY = 0 - centerY;
+        const maxX = width - centerX;
+        const maxY = height - centerY;
         this._pixiObject.clear();
         this._pixiObject.beginFill(0x0033ff);
         this._pixiObject.lineStyle(1, 0xffd900, 1);
-        this._pixiObject.moveTo(-width / 2, -height / 2);
-        this._pixiObject.lineTo(width / 2, -height / 2);
-        this._pixiObject.lineTo(width / 2, height / 2);
-        this._pixiObject.lineTo(-width / 2, height / 2);
+        this._pixiObject.moveTo(minX, minY);
+        this._pixiObject.lineTo(maxX, minY);
+        this._pixiObject.lineTo(maxX, maxY);
+        this._pixiObject.lineTo(minX, maxY);
         this._pixiObject.endFill();
 
-        this._pixiObject.moveTo(-width / 2, -height / 2);
-        this._pixiObject.lineTo(width / 2, height / 2);
-        this._pixiObject.moveTo(width / 2, -height / 2);
-        this._pixiObject.lineTo(-width / 2, height / 2);
+        this._pixiObject.moveTo(minX, minY);
+        this._pixiObject.lineTo(maxX, maxY);
+        this._pixiObject.moveTo(maxX, minY);
+        this._pixiObject.lineTo(minX, maxY);
 
-        this._pixiObject.position.x = this._instance.getX() + width / 2;
-        this._pixiObject.position.y = this._instance.getY() + height / 2;
+        const originPoint = this.getOriginPoint();
+        this._pixiObject.position.x =
+          this._instance.getX() - width * (originPoint[0] - centerPoint[0]);
+        this._pixiObject.position.y =
+          this._instance.getY() - height * (originPoint[1] - centerPoint[1]);
         this._pixiObject.angle = this._instance.getAngle();
       }
 
```

Comment:
should we also add a getDefaultDepth() function? (mirrorring what's done for other models)

Reply:
I'm wondering if this is useful when drag an object on the scene and it creates the instance, to ensure the depth is properly set for the instance. (I fixed this issue for the 3D Cube in 2D)

Reply:
Good catch